File: app/src/main/java/xyz/hetula/homefy/player/HomefyPlayer.java
/*
 * MIT License
 *
 * Copyright (c) 2017 Tuomo Heino
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package xyz.hetula.homefy.player;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.media.AudioAttributes;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.net.Uri;
import android.net.wifi.WifiManager;
import android.os.Handler;
import android.os.PowerManager;
import android.support.annotation.Nullable;
import android.support.v4.media.session.MediaControllerCompat;
import android.support.v4.media.session.MediaSessionCompat;
import android.support.v4.media.session.PlaybackStateCompat;
import android.util.Log;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.TimeUnit;

import xyz.hetula.homefy.service.Homefy;

import static android.media.AudioManager.AUDIOFOCUS_LOSS_TRANSIENT;
import static android.media.AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK;

public class HomefyPlayer {
    private static final String TAG = "HomefyPlayer";

    private BecomingNoisyReceiver myNoisyAudioStreamReceiver = new BecomingNoisyReceiver();
    private Set<PlaybackListener> mPlaybackListeners = new HashSet<>();
    private Handler mHandler = new Handler();

    private final List<Song> mPlaylist;

    private Context mContext;
    private MediaPlayer mPlayer;
    private MediaSessionCompat mSession;
    private MediaControllerCompat mController;
    private PlaybackStateCompat.Builder mStateBuilder;
    private AudioManager.OnAudioFocusChangeListener afChangeListener = this::onAudioFocusChange;
    private WifiManager.WifiLock mWifiLock;

    private Song mNowPlaying;

    public HomefyPlayer(Context context) {
        mPlaylist = new ArrayList<>();

        mContext = context;
        IntentFilter intentFilter = new IntentFilter(AudioManager.ACTION_AUDIO_BECOMING_NOISY);
        context.registerReceiver(myNoisyAudioStreamReceiver, intentFilter);

        mSession = new MediaSessionCompat(context, "Homefy Player");

        mSession.setFlags(MediaSessionCompat.FLAG_HANDLES_MEDIA_BUTTONS |
                MediaSessionCompat.FLAG_HANDLES_TRANSPORT_CONTROLS);

        mStateBuilder = new PlaybackStateCompat.Builder().setActions(
                PlaybackStateCompat.ACTION_PLAY |
                        PlaybackStateCompat.ACTION_PLAY_PAUSE);
        mSession.setPlaybackState(mStateBuilder.build());
        mSession.setActive(true);

        mController = mSession.getController();

        mPlayer = new MediaPlayer();
        mPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC);
        mPlayer.setAudioAttributes(new AudioAttributes.Builder()
                .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                .setUsage(AudioAttributes.USAGE_MEDIA)
                .build());
        mPlayer.setOnCompletionListener(this::onPlayComplete);
        mPlayer.setOnPreparedListener(this::onPrepareComplete);
        mPlayer.setOnErrorListener(this::onError);
        mPlayer.setOnBufferingUpdateListener(this::onBuffering);
        mPlayer.setWakeMode(mContext.getApplicationContext(), PowerManager.PARTIAL_WAKE_LOCK);

        mWifiLock = ((WifiManager) mContext
                .getApplicationContext().getSystemService(Context.WIFI_SERVICE))
                .createWifiLock(WifiManager.WIFI_MODE_FULL, "HomefyPlayerWifiLock");

        MediaSessionCompat.Callback callback = new MediaSessionCompat.Callback() {
            @Override
            public void onPlay() {
                AudioManager am = (AudioManager) mContext.getSystemService(Context.AUDIO_SERVICE);
                pauseResume();
            }

            @Override
            public void onPause() {
                AudioManager am = (AudioManager) mContext.getSystemService(Context.AUDIO_SERVICE);
                pauseResume();
            }

            @Override
            public void onSkipToNext() {
                next();
            }

            @Override
            public void onSkipToPrevious() {
                previous();
            }

            @Override
            public void onStop() {
                AudioManager am = (AudioManager) mContext.getSystemService(Context.AUDIO_SERVICE);
                stop();
            }
        };
    }

    public MediaSessionCompat getSession() {
        return mSession;
    }

    public void release() {
        mPlaybackListeners.clear();

        mSession.release();
        mSession = null;

        mPlayer.release();
        mPlayer = null;

        if (mWifiLock.isHeld()) {
            mWifiLock.release();
        }

        // Just in case as Android does not behave nicely when unregistering non-registered
        // receivers. Should never happen, but it is better to catch it than crash whole app.
        try {
            mContext.unregisterReceiver(myNoisyAudioStreamReceiver);
        } catch (IllegalStateException ex) {
            Log.w(TAG, "Releasing unregistered Noisy Receiver!", ex);
        }
        myNoisyAudioStreamReceiver = null;
        mContext = null;
    }

    @Nullable
    public Song nowPlaying() {
        return mNowPlaying;
    }

    public void play(Song song, @Nullable List<Song> playlist) {
        if (setupPlay(song) && playlist != null) {
            mPlaylist.clear();
            mPlaylist.addAll(playlist);
        }
    }

    public void pauseResume() {
        if (mNowPlaying == null) return;
        if (mPlayer.isPlaying()) {
            mPlayer.pause();
            for (PlaybackListener listener : mPlaybackListeners) {
                listener.onSongPlay(nowPlaying(), PlaybackListener.STATE_PAUSE);
            }
        } else {
            mPlayer.start();
            for (PlaybackListener listener : mPlaybackListeners) {
                listener.onSongPlay(nowPlaying(), PlaybackListener.STATE_RESUME);
            }
        }
    }

    public boolean isPlaying() {
        return mNowPlaying != null && mPlayer.isPlaying();
    }

    public boolean isPaused() {
        return mNowPlaying != null && !mPlayer.isPlaying();
    }

    public void stop() {
        abandonAudioFocus();
        for (PlaybackListener listener : mPlaybackListeners) {
            listener.onSongPlay(nowPlaying(), PlaybackListener.STATE_STOP);
        }
        mPlayer.stop();
    }

    public void previous() {
    }

    public void next() {
        if (mPlaylist.isEmpty()) return;
        // Normal Playback mode implementation
        // This needs own classes/methods etc for proper impl
        // TODO Implement Playback modes
        int id = mPlaylist.indexOf(mNowPlaying);
        if (id == -1) return;
        id++;
        if (id >= mPlaylist.size()) return;
        setupPlay(mPlaylist.get(id));
    }

    private boolean setupPlay(Song song) {
        AudioManager am = (AudioManager) mContext.getSystemService(Context.AUDIO_SERVICE);
        // Request audio focus for playback
        int result = am.requestAudioFocus(afChangeListener,
                AudioManager.STREAM_MUSIC,
                AudioManager.AUDIOFOCUS_GAIN);

        if (result != AudioManager.AUDIOFOCUS_REQUEST_GRANTED) {
            Log.w(TAG, "No AudioFocus Granted!");
            return false;
        }
        try {
            mWifiLock.acquire();
            Uri uri = Uri.parse(Homefy.library().getPlayPath(song));
            mPlayer.reset();
            Map<String, String> headers = new HashMap<>();
            Homefy.protocol().addAuthHeader(headers);
            mPlayer.setDataSource(mContext, uri, headers);
            mPlayer.prepareAsync();

            mNowPlaying = song;
            return true;
        } catch (IOException e) {
            Log.e(TAG, "Error when playing", e);
            return false;
        }
    }

    private void onPrepareComplete(MediaPlayer mp) {
        mp.start();
        for (PlaybackListener listener : mPlaybackListeners) {
            listener.onSongPlay(nowPlaying(), PlaybackListener.STATE_PLAY);
        }
    }

    private void onPlayComplete(MediaPlayer mp) {
        abandonAudioFocus();
        next();
    }

    private boolean onError(MediaPlayer mp, int what, int extra) {
        Log.e(TAG, "Playback Error: " + what + " extra: " + extra);
        abandonAudioFocus();
        for (PlaybackListener listener : mPlaybackListeners) {
            listener.onSongPlay(nowPlaying(), PlaybackListener.STATE_STOP);
        }
        return true;
    }

    private void onBuffering(MediaPlayer mediaPlayer, int i) {
        if (i < 100) return;
        Log.d(TAG, "Fully buffered!");
        if (mWifiLock.isHeld()) {
            mWifiLock.release();
        }
    }

    private void abandonAudioFocus() {
        AudioManager am = (AudioManager) mContext.getSystemService(Context.AUDIO_SERVICE);
        am.abandonAudioFocus(afChangeListener);
        if (mWifiLock.isHeld()) {
            mWifiLock.release();
        }
    }

    private void onAudioFocusChange(int focusChange) {
        if (focusChange == AudioManager.AUDIOFOCUS_LOSS) {
            mController.getTransportControls().pause();
            mHandler.postDelayed(mController.getTransportControls()::stop,
                    TimeUnit.SECONDS.toMillis(30));
        } else if (focusChange == AUDIOFOCUS_LOSS_TRANSIENT) {
            mController.getTransportControls().pause();
        } else if (focusChange == AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK) {
            mPlayer.setVolume(0.1f, 0.1f);
        } else if (focusChange == AudioManager.AUDIOFOCUS_GAIN) {
            mPlayer.setVolume(1, 1);
            // TODO deside if continueing playback is wanted here.
            // ducking wont pause, so receiving messages doesn't cause problems
            // Phone calls etc can cause problems if headset is removed during call
            // after call end playback can possibly continue from wrong speaker.
        }
    }

    public void unregisterPlaybackListener(PlaybackListener mPlaybackListener) {
        mPlaybackListeners.remove(mPlaybackListener);
    }

    public void registerPlaybackListener(PlaybackListener mPlaybackListener) {
        mPlaybackListeners.add(mPlaybackListener);
    }

    private class BecomingNoisyReceiver extends BroadcastReceiver {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (AudioManager.ACTION_AUDIO_BECOMING_NOISY.equals(intent.getAction())) {
                if (mNowPlaying == null) return;
                mPlayer.pause();
            }
        }
    }
}

File: app/src/main/kotlin/xyz.hetula.homefy/player/HomefyPlayer.kt
/*
 * Copyright (c) 2018 Tuomo Heino
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package xyz.hetula.homefy.player

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.media.*
import android.media.AudioManager.*
import android.net.Uri
import android.net.wifi.WifiManager
import android.os.AsyncTask
import android.os.PowerManager
import android.os.SystemClock
import android.support.v4.media.session.MediaControllerCompat
import android.support.v4.media.session.MediaSessionCompat
import android.support.v4.media.session.PlaybackStateCompat
import android.util.Log
import xyz.hetula.homefy.library.HomefyLibrary
import xyz.hetula.homefy.service.protocol.HomefyProtocol
import java.io.ByteArrayInputStream
import java.io.IOException
import java.util.*
import kotlin.collections.ArrayList


/**
 * @author Tuomo Heino
 * @version 1.0
 * @since 1.0
 */
open class HomefyPlayer(private val mProtocol: HomefyProtocol,
                        private val mLibrary: HomefyLibrary) {
    private val playNextThresholdMs = 500L
    private val mPlaybackListeners = HashSet<(Song?, Int, Int) -> Unit>()
    private val mPlayback = Playback()

    private var mNoisyAudioStreamReceiver: BecomingNoisyReceiver? = null
    private var mMediaSession: MediaSessionCompat? = null
    private var mController: MediaControllerCompat? = null
    private var mWifiLock: WifiManager.WifiLock? = null
    private var mPlayer: MediaPlayer? = null
    private var mContext: Context? = null

    private var afChangeListener = { focusChange: Int -> onAudioFocusChange(focusChange) }
    private var mLastPlayPress = 0L
    private var mHasFocus = false
    private var mAudioFocusRequest: AudioFocusRequest? = null
    private var showAlbumArtInLockScreen = false

    fun initalize(context: Context): MediaSessionCompat {
        Log.d(TAG, "initialize: Initializing!")
        val mediaSession = MediaSessionCompat(context, "Homefy Player")
        mediaSession.setFlags(MediaSessionCompat.FLAG_HANDLES_MEDIA_BUTTONS or
                MediaSessionCompat.FLAG_HANDLES_TRANSPORT_CONTROLS)
        mediaSession.setCallback(object : MediaSessionCompat.Callback() {
            override fun onPlay() {
                pauseResume()
            }

            override fun onPause() {
                pauseResume()
            }

            override fun onSkipToNext() {
                next()
            }

            override fun onSkipToPrevious() {
                previous()
            }

            override fun onStop() {
                stop()
            }
        })

        val player = MediaPlayer()
        player.setAudioAttributes(AudioAttributes.Builder()
                .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                .setUsage(AudioAttributes.USAGE_MEDIA)
                .build())

        player.setOnCompletionListener(this::onPlayComplete)
        player.setOnPreparedListener(this::onPrepareComplete)
        player.setOnErrorListener(this::onError)
        player.setOnBufferingUpdateListener(this::onBuffering)
        player.setWakeMode(context.applicationContext, PowerManager.PARTIAL_WAKE_LOCK)

        mWifiLock = (context.applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager)
                .createWifiLock(WifiManager.WIFI_MODE_FULL, "HomefyPlayerWifiLock")

        mContext = context
        mPlayer = player
        mMediaSession = mediaSession
        mController = mediaSession.controller

        val noisyAudioStreamReceiver = BecomingNoisyReceiver()
        val intentFilter = IntentFilter(AudioManager.ACTION_AUDIO_BECOMING_NOISY)
        context.registerReceiver(noisyAudioStreamReceiver, intentFilter)
        mNoisyAudioStreamReceiver = noisyAudioStreamReceiver

        mediaSession.isActive = true
        updatePlaybackState(PlaybackStateCompat.STATE_NONE)

        return mediaSession
    }

    fun release(context: Context) {
        releaseBroadcastReceiver(context, mNoisyAudioStreamReceiver)
        mNoisyAudioStreamReceiver = null

        loseAudioFocus()
        mPlaybackListeners.clear()

        mPlayer?.release()
        mPlayer = null

        mMediaSession?.release()
        mMediaSession = null
        mController = null

        releaseWifiLock(mWifiLock)
        mWifiLock = null

        mContext = null
    }

    fun nowPlaying(): Song? {
        return mPlayback.getCurrent()
    }

    fun queue(song: Song) {
        mPlayback.queueSong(Collections.singletonList(song))
    }

    fun queue(song: List<Song>) {
        mPlayback.queueSong(song)
    }

    fun play(song: Song, playlist: ArrayList<Song>?) {
        val player = mPlayer ?: return
        if (setupPlay(player, song)) {
            mPlayback.getCurrent()?.clearAlbumArt()
            if (playlist == null)
                mPlayback.playSong(song, ArrayList())
            else
                mPlayback.playSong(song, playlist)
        }
    }

    fun pauseResume() {
        if (mPlayback.hasPlayback()) {
            return
        }
        if (shouldPlayNext()) {
            next()
            return
        }
        mLastPlayPress = SystemClock.elapsedRealtime()

        player {
            if (it.isPlaying) {
                pause()
            } else {
                play()
            }
        }
    }

    private fun shouldPlayNext(): Boolean {
        if (mLastPlayPress == 0L) {
            return false
        }
        val diff = SystemClock.elapsedRealtime() - mLastPlayPress
        return diff < playNextThresholdMs
    }

    private fun pause() {
        player { mediaPlayer ->
            if (mPlayback.hasPlayback() || !mediaPlayer.isPlaying) return
            mediaPlayer.pause()
            mPlaybackListeners.forEach { it(nowPlaying()!!, STATE_PAUSE, -1) }
            updatePlaybackState(PlaybackStateCompat.STATE_PAUSED)
        }
    }

    private fun play() {
        if (mPlayback.hasPlayback()) return
        player { mediaPlayer ->
            mediaPlayer.start()
            mPlaybackListeners.forEach { it(nowPlaying()!!, STATE_RESUME, -1) }
            updatePlaybackState(PlaybackStateCompat.STATE_PLAYING)
        }
    }

    val isPlaying: Boolean
        get(): Boolean {
            val player = mPlayer ?: return false
            return !mPlayback.hasPlayback() && player.isPlaying
        }

    val isPaused: Boolean
        get(): Boolean {
            val player = mPlayer ?: return false
            return !mPlayback.hasPlayback() && !player.isPlaying
        }

    fun stop() {
        mPlaybackListeners.forEach { it(nowPlaying()!!, STATE_STOP, -1) }
        mPlayback.getCurrent()?.clearAlbumArt()
        mPlayback.stop()
        mPlayer?.stop()
        updatePlaybackState(PlaybackStateCompat.STATE_STOPPED)
    }

    fun previous() {
        mPlayback.getCurrent()?.clearAlbumArt()
        mPlayback.previous()
        if (mPlayback.hasPlayback()) return
        updatePlaybackState(PlaybackStateCompat.STATE_SKIPPING_TO_PREVIOUS)
        mPlayback.getCurrent { song ->
            player { setupPlay(it, song) }
        }
    }

    fun next() {
        mPlayback.getCurrent()?.clearAlbumArt()
        mPlayback.next()
        if (mPlayback.hasPlayback()) return
        updatePlaybackState(PlaybackStateCompat.STATE_SKIPPING_TO_NEXT)
        mPlayback.getCurrent { song ->
            player { setupPlay(it, song) }
        }
    }

    fun cyclePlaybackMode(): PlaybackMode {
        mPlayback.cyclePlaybackMode()
        return mPlayback.playbackMode
    }

    fun queryPosition(): Int {
        if (mPlayback.hasPlayback()) return 0
        val player = mPlayer ?: return 0
        return player.currentPosition / 1000
    }

    fun queryPositionMs(): Int {
        if (mPlayback.hasPlayback()) return 0
        val player = mPlayer ?: return 0
        return player.currentPosition
    }

    fun seekTo(position: Int) {
        mPlayer?.seekTo(position * 1000)
    }

    private fun queryPosMs(): Long {
        if (mPlayback.hasPlayback()) return 0L
        val player = mPlayer ?: return 0L
        return player.currentPosition.toLong()
    }

    private fun setupPlay(player: MediaPlayer, song: Song): Boolean {
        if (!mHasFocus) {
            tryGainAudioFocus()
            if (!mHasFocus) {
                return false
            }
        }
        Log.d(TAG, "Setuping new song to play!")
        return try {
            acquireWifiLock(mWifiLock)
            val uri = Uri.parse(mLibrary.getPlayPath(song))
            player.reset()
            val headers = HashMap<String, String>()
            mProtocol.addAuthHeader(headers)
            player.setDataSource(mContext!!, uri, headers)
            mediaSession { it.setMetadata(song.toMediaMetadata(showAlbumArtInLockScreen)) }

            AlbumArtRetriever(song, uri, headers) { originalSong, bitmap ->
                Log.d(TAG, "AlbumArtRetriever: A result! $bitmap")
                val now = nowPlaying() ?: return@AlbumArtRetriever
                if (now.id != originalSong.id) {
                    return@AlbumArtRetriever
                }
                mediaSession { mediaSession ->
                    Log.d(TAG, "AlbumArtRetriever: Setting new Metadata!")
                    song.albumArt = bitmap
                    mediaSession.setMetadata(song.toMediaMetadata(showAlbumArtInLockScreen))
                    mPlaybackListeners.forEach { it(song, STATE_PLAY, -1) }
                }
            }.execute()
            updatePlaybackState(PlaybackStateCompat.STATE_CONNECTING)
            player.prepareAsync()
            true
        } catch (e: IOException) {
            Log.e(TAG, "Error when playing", e)
            releaseWifiLock(mWifiLock)
            false
        }
    }

    private fun onPrepareComplete(mp: MediaPlayer) {
        Log.d(TAG, "onPrepareComlete, Starting Playback")
        mp.start()
        mPlaybackListeners.forEach { it(nowPlaying()!!, STATE_PLAY, -1) }
        updatePlaybackState(PlaybackStateCompat.STATE_PLAYING)
    }

    private fun onPlayComplete(mp: MediaPlayer) {
        Log.d(TAG, "onPlayComplete $mp")
        next()
    }

    private fun onError(mp: MediaPlayer, what: Int, extra: Int): Boolean {
        Log.e(TAG, "Playback Error: $what extra: $extra MediaPlayer: $mp")
        MediaPlayer.MEDIA_ERROR_IO
        mPlaybackListeners.forEach { it(nowPlaying(), STATE_STOP, -1) }
        updatePlaybackState(PlaybackStateCompat.STATE_ERROR)
        return true
    }

    private fun onBuffering(mediaPlayer: MediaPlayer, i: Int) {
        mPlaybackListeners.forEach { it(nowPlaying(), STATE_BUFFERING, i) }
        if (i < 100) return
        Log.d(TAG, "Fully buffered! $mediaPlayer")
        releaseWifiLock(mWifiLock)
    }

    private fun tryGainAudioFocus() {
        if (mHasFocus) {
            loseAudioFocus()
        }
        context {
            val am = it.getSystemService(Context.AUDIO_SERVICE) as AudioManager

            val req = AudioFocusRequest.Builder(AUDIOFOCUS_GAIN)
                    .setOnAudioFocusChangeListener(afChangeListener)
                    .setAcceptsDelayedFocusGain(false)
                    .setAudioAttributes(AudioAttributes.Builder()
                            .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                            .setUsage(AudioAttributes.USAGE_MEDIA)
                            .build())
                    .build()
            val result = am.requestAudioFocus(req)


            mHasFocus =
                    if (result != AudioManager.AUDIOFOCUS_REQUEST_GRANTED) {
                        Log.w(TAG, "No AudioFocus Granted!")
                        false
                    } else {
                        true
                    }
        }
    }

    private fun loseAudioFocus() {
        context {
            val am = it.getSystemService(Context.AUDIO_SERVICE) as AudioManager
            val req = mAudioFocusRequest ?: return
            am.abandonAudioFocusRequest(req)
        }
    }

    private fun onAudioFocusChange(focusChange: Int) {
        Log.d(TAG, "Focus Change!! $focusChange")
        when (focusChange) {
            AudioManager.AUDIOFOCUS_LOSS -> pause()
            AUDIOFOCUS_LOSS_TRANSIENT -> pause()
            AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK -> player { it.setVolume(0.1f, 0.1f) }
            AudioManager.AUDIOFOCUS_GAIN -> player { it.setVolume(1f, 1f) }
        }
    }

    fun unregisterPlaybackListener(mPlaybackListener: (Song?, Int, Int) -> Unit) {
        mPlaybackListeners.remove(mPlaybackListener)
    }

    fun registerPlaybackListener(mPlaybackListener: (Song?, Int, Int) -> Unit) {
        mPlaybackListeners.add(mPlaybackListener)
    }

    private fun updatePlaybackState(state: Int) {
        val stateBuilder = PlaybackStateCompat.Builder().setActions(
                PlaybackStateCompat.ACTION_PLAY or
                        PlaybackStateCompat.ACTION_PAUSE or
                        PlaybackStateCompat.ACTION_PLAY_PAUSE or
                        PlaybackStateCompat.ACTION_SKIP_TO_NEXT or
                        PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS or
                        PlaybackStateCompat.ACTION_STOP)

        stateBuilder.setState(state, queryPosMs(), 1f)
        mediaSession { it.setPlaybackState(stateBuilder.build()) }
    }

    private fun releaseBroadcastReceiver(context: Context,
                                         noiseAudioStreamReceiver: BroadcastReceiver?) {
        noiseAudioStreamReceiver ?: return
        try {
            context.unregisterReceiver(noiseAudioStreamReceiver)
        } catch (ex: IllegalStateException) {
            Log.w(TAG, "Can't release: " + noiseAudioStreamReceiver.javaClass.simpleName, ex)
        }
    }

    private fun acquireWifiLock(wifiLock: WifiManager.WifiLock?) {
        wifiLock ?: return
        if (!wifiLock.isHeld) {
            wifiLock.acquire()
        }
    }

    private fun releaseWifiLock(wifiLock: WifiManager.WifiLock?) {
        wifiLock ?: return
        if (wifiLock.isHeld) {
            wifiLock.release()
        }
    }

    private inline fun mediaSession(mediaSessionCallback: (MediaSessionCompat) -> Unit) {
        val mediaSession = mMediaSession
        if (mediaSession == null) {
            Log.w(TAG, "MediaSession: No MediaSession instance!!")
            return
        }
        mediaSessionCallback(mediaSession)
    }

    private inline fun player(playerCallback: (MediaPlayer) -> Unit) {
        val player = mPlayer
        if (player == null) {
            Log.w(TAG, "MediaPlayer: No Player instance!!")
            return
        }
        playerCallback(player)
    }

    private inline fun context(contextCallback: (Context) -> Unit) {
        val context = mContext
        if (context == null) {
            Log.w(TAG, "Context: No Context set!")
            return
        }
        contextCallback(context)
    }

    private inner class BecomingNoisyReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            if (AudioManager.ACTION_AUDIO_BECOMING_NOISY == intent.action) {
                pause()
            }
        }
    }

    private class AlbumArtRetriever(val song: Song,
                                    val uri: Uri,
                                    val headers: MutableMap<String, String>,
                                    val albumArtCallback: (Song, Bitmap) -> Unit) :
            AsyncTask<Void, Void, Bitmap?>() {

        override fun doInBackground(vararg params: Void?): Bitmap? {
            val mmr = MediaMetadataRetriever()
            try {
                Log.v(TAG, "AlbumArtRetriever: $uri")
                mmr.setDataSource(uri.toString(), headers)
                val artBytes = mmr.embeddedPicture ?: return null
                val inputStream = ByteArrayInputStream(artBytes)
                val bitmap = BitmapFactory.decodeStream(inputStream)
                inputStream.close()
                return bitmap
            } catch (ex: RuntimeException) {
                Log.w(TAG, "AlbumArtRetriever: $song", ex)
                return null
            } finally {
                mmr.release()
            }
        }

        override fun onPostExecute(result: Bitmap?) {
            result ?: return
            albumArtCallback(song, result)
        }
    }

    companion object {
        private const val TAG = "HomefyPlayer"

        /*
         * Playback Codes
         */
        const val STATE_PLAY = 0
        const val STATE_PAUSE = 1
        const val STATE_RESUME = 2
        const val STATE_STOP = 3
        const val STATE_BUFFERING = 4
    }
}
